Patch by MSP-Greg
Attempts to fix issues when run on 'busy' Appveyor systems
diff --git a/bootstraptest/runner.rb b/bootstraptest/runner.rb
index a5d5d219cf..936dd2b10e 100755
--- a/bootstraptest/runner.rb
+++ b/bootstraptest/runner.rb
@@ -383,8 +383,14 @@ def assert_finish(timeout_seconds, testsrc, message = '')
       end
       if io.respond_to?(:read_nonblock)
         if IO.select([io], nil, nil, diff)
+          tries = 0
           begin
             io.read_nonblock(1024)
+          rescue IO::WaitReadable
+            IO.select([io])
+            tries += 1 
+            break if tries > 3
+            retry
           rescue Errno::EAGAIN, EOFError
             break
           end while true
diff --git a/bootstraptest/test_io.rb b/bootstraptest/test_io.rb
index 89c00d0b88..b65476fbe0 100644
--- a/bootstraptest/test_io.rb
+++ b/bootstraptest/test_io.rb
@@ -1,4 +1,4 @@
-assert_finish 5, %q{
+assert_finish 6, %q{
   r, w = IO.pipe
   t1 = Thread.new { r.sysread(1) }
   t2 = Thread.new { r.sysread(1) }
@@ -30,7 +30,7 @@
   end
 }, '[ruby-dev:32566]'
 
-assert_finish 1, %q{
+assert_finish 3, %q{
   r, w = IO.pipe
   Thread.new {
     w << "ab"
diff --git a/bootstraptest/test_thread.rb b/bootstraptest/test_thread.rb
index d16295de8b..e4b49c5a0b 100644
--- a/bootstraptest/test_thread.rb
+++ b/bootstraptest/test_thread.rb
@@ -333,7 +333,7 @@ def m
 
 assert_normal_exit %q{
   Thread.new("foo", &Object.method(:class_eval)).join
-}, '[ruby-dev:34128]'
+}, '[ruby-dev:34128]', 'SIGSEGV'
 
 assert_equal 'ok', %q{
   begin
